src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/tools/cortisol/IngestTask.java

```diff
@@ -0,0 +1,172 @@
+    }
+
+    private void createIndex(final String mappingTemplateJson, int nPrimaryShards, int nReplicaShards) throws IOException {
+        final CreateIndexRequest cir = new CreateIndexRequest(indexName);
+        cir.mapping(mappingTemplateJson, XContentType.JSON);
+        cir.settings(CortisolHelper.buildIndexSettingsJson(nPrimaryShards, nReplicaShards), XContentType.JSON);
+        final CreateIndexResponse response = client.indices().create(cir, RequestOptions.DEFAULT);
+        assert response.isAcknowledged();
+    }
+
+    private void makeBulkRequest(final List<String> bulkDocs) {
+        final BulkRequest bulkRequest = new BulkRequest();
+        for (String bulkDoc : bulkDocs) {
+            final IndexRequest indexRequest = Requests.indexRequest(indexName);
+            indexRequest.source(bulkDoc, XContentType.JSON);
+            bulkRequest.add(indexRequest);
+        }
+        client.bulkAsync(bulkRequest, RequestOptions.DEFAULT, new ActionListener<BulkResponse>() {
+            @Override
+            public void onResponse(BulkResponse bulkItemResponses) {
+                if (!bulkItemResponses.hasFailures()) {
+                    successCount.addAndGet(bulkDocs.size());
+                } else {
+                    for (BulkItemResponse response : bulkItemResponses) {
+                        if (response.isFailed()) {
+                            failureCount.incrementAndGet();
+                        } else {
+                            successCount.incrementAndGet();
+                        }
+                    }
+                }
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                failureCount.addAndGet(bulkDocs.size());
+            }
+        });
+        requestCount.addAndGet(bulkDocs.size());
+    }
+
+    private class TokenBucket {
+        private final Lock lock = new ReentrantLock();
+        private final Condition hasTokens = lock.newCondition();
+        private final ScheduledExecutorService ses = Executors.newSingleThreadScheduledExecutor();
+
+        private final int maxTokens;
+        private int nTokens;
+
+        public TokenBucket(final int maxTokens) {
+            this.maxTokens = maxTokens;
+            this.ses.schedule(this::refill, 1, TimeUnit.SECONDS);
+        }
+
+        public int take() {
+            lock.lock();
+            try {
+                while (nTokens == 0) {
+                    hasTokens.await();
+                }
+                return nTokens--;
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            } finally {
+                lock.unlock();
+            }
+
+            return 0;
+        }
+
+        private void refill() {
+            lock.lock();
+            try {
+                nTokens = maxTokens;
+                hasTokens.signal();
+            } finally {
+                lock.unlock();
+            }
```

Comment:
Why doesn't the take() method just replenish the bucket by 1 once it's done?

Reply:
The work done after taking a token is dependent on the time each bulk() takes. We won't be able to guarantee a constant request rate. Some more explanation is in the next comment.